File: day1-kotlin-syntax-crash-course/java/ValVarExample.java
/**
 * Day 1 - val/var Equivalent in Java
 * 
 * Java uses:
 *   - final → immutable (like Kotlin's val)
 *   - non-final → mutable (like Kotlin's var)
 *   - No local variable type inference before Java 10
 */
public class ValVarExample {

    // Java class-level fields: ALWAYS need explicit type (no inference at class level)
    private static final String NAME = "Kotlin";    // final = immutable (like val)
    private static int counter = 0;                  // non-final = mutable (like var)

    public static void main(String[] args) {

        // ═══════════════════════════════════════════════════════════════
        // 1. FINAL vs NON-FINAL (val vs var equivalent)
        // ═══════════════════════════════════════════════════════════════

        // final variable (like Kotlin's val) — cannot be reassigned
        final String language = "Kotlin";
        // language = "Java";  // ❌ COMPILE ERROR: cannot assign final variable

        // non-final variable (like Kotlin's var) — can be reassigned
        int count = 0;
        count = 1;  // ✅ OK
        count = 2;  // ✅ OK

        // ═══════════════════════════════════════════════════════════════
        // 2. TYPE INFERENCE with var (Java 10+)
        // ═══════════════════════════════════════════════════════════════

        // Java 10+ local variable type inference (like Kotlin's type inference)
        var message = "Hello";          // inferred as String
        var number = 42;                // inferred as int
        var decimal = 3.14;             // inferred as double
        var list = java.util.List.of(1, 2, 3);  // inferred as List<Integer>

        System.out.println("message type: " + message.getClass().getSimpleName());
        System.out.println("number type: " + Integer.TYPE.getSimpleName());

        // ═══════════════════════════════════════════════════════════════
        // 3. COMPARISON TABLE
        // ═══════════════════════════════════════════════════════════════
        //
        //   Kotlin          Java 17+              Notes
        //   ──────────────  ────────────────────  ─────────────────────────
        //   val x = 5       final int x = 5       Immutable (reassignment forbidden)
        //   var y = 5       int y = 5              Mutable (reassignment allowed)
        //   val z: Int      final int z            Explicit type
        //   var a: Int = 5  int a = 5              Explicit type
        //   var b = 5       var b = 5 (Java 10+)  Type inference (local only in Java)
        //
        // ═══════════════════════════════════════════════════════════════

        System.out.println("language (final): " + language);
        System.out.println("count (mutable): " + count);
        System.out.println("message (var inferred): " + message);

        // ═══════════════════════════════════════════════════════════════
        // 4. KEY DIFFERENCES
        // ═══════════════════════════════════════════════════════════════
        //
        // DIFFERENCE 1: Java var is LOCAL-ONLY, Kotlin var works everywhere
        //   - Java: var only works inside methods, NOT for fields
        //   - Kotlin: val/var works for both properties and local variables
        //
        // DIFFERENCE 2: Java requires explicit types for class fields
        //   - Java: private String name = "test";  // MUST specify String
        //   - Kotlin: val name = "test"            // Inferred as String at property level
        //
        // DIFFERENCE 3: Kotlin val means REFERENCE immutability, not deep immutability
        //   - val list = mutableListOf(1, 2, 3)
        //   - list.add(4)  // ✅ OK! List content changed, reference didn't
        //
        // DIFFERENCE 4: Java final on对象引用 = same behavior as Kotlin val
        //   - final var list = new ArrayList<>();
        //   - list.add(1);  // ✅ OK! Same as Kotlin val
        //
        // ═══════════════════════════════════════════════════════════════

        // Example: val does NOT prevent content mutation
        final var mutableList = new java.util.ArrayList<>(java.util.List.of(1, 2, 3));
        mutableList.add(4);  // ✅ OK — reference is final, but content is mutable
        System.out.println("mutableList after add: " + mutableList);

        // Example: const val equivalent
        // Kotlin: const val MAX_SIZE = 100  (compile-time constant)
        // Java: static final int MAX_SIZE = 100;  (also compile-time constant)
        // Both must be primitives or String, both inlined at compile time

        // ═══════════════════════════════════════════════════════════════
        // 5. SUMMARY FOR JAVA → KOTLIN MIGRATION
        // ═══════════════════════════════════════════════════════════════
        //
        // Java                          →  Kotlin
        // ─────────────────────────────────────────────────
        // final int x = 5;              →  val x = 5
        // int x = 5;                    →  var x = 5
        // private final String s = "";  →  private val s = ""
        // private String s = "";        →  private var s = ""
        // var x = 5; (Java 10+)        →  var x = 5 (same!)
        // static final int C = 10;     →  const val C = 10
        //
        // ═══════════════════════════════════════════════════════════════
    }
}

File: day1-kotlin-syntax-crash-course/kotlin/ValVarExample.kt
/**
 * Day 1 - val/var in Kotlin
 *
 * Kotlin uses:
 *   - val → immutable reference (read-only, like Java's final)
 *   - var → mutable reference (re-assignable)
 *   - Type inference works everywhere (properties AND local variables)
 */

fun main() {

    // ═══════════════════════════════════════════════════════════════
    // 1. val (immutable — like Java's final)
    // ═══════════════════════════════════════════════════════════════

    val language = "Kotlin"    // Type inferred as String
    // language = "Java"       // ❌ COMPILE ERROR: Val cannot be reassigned

    val number: Int = 42      // Explicit type (type after variable name!)
    val decimal = 3.14        // Inferred as Double
    val isValid = true        // Inferred as Boolean

    // ═══════════════════════════════════════════════════════════════
    // 2. var (mutable — like Java's non-final)
    // ═══════════════════════════════════════════════════════════════

    var count = 0
    count = 1    // ✅ OK
    count = 2    // ✅ OK

    var name: String = "Alice"
    name = "Bob"  // ✅ OK

    // ═══════════════════════════════════════════════════════════════
    // 3. TYPE INFERENCE — Kotlin does it everywhere
    // ═══════════════════════════════════════════════════════════════

    // Local variables — type inferred
    val inferredString = "Hello"       // String
    val inferredInt = 42               // Int
    val inferredDouble = 3.14          // Double
    val inferredList = listOf(1, 2, 3) // List<Int>

    // Class properties — type inferred (Java CANNOT do this!)
    // In Kotlin, even properties get type inference
    // Java: private String name = "test";  // MUST specify String
    // Kotlin: private val name = "test"    // Inferred

    println("language: $language, count: $count")

    // ═══════════════════════════════════════════════════════════════
    // 4. val = REFERENCE immutability, NOT deep immutability
    // ═══════════════════════════════════════════════════════════════

    val list = mutableListOf(1, 2, 3)
    list.add(4)          // ✅ OK! List content changed, reference didn't
    println("list: $list")  // [1, 2, 3, 4]

    // To make it truly immutable:
    val immutableList = listOf(1, 2, 3)
    // immutableList.add(4)  // ❌ COMPILE ERROR: listOf returns read-only List

    // ═══════════════════════════════════════════════════════════════
    // 5. const val (compile-time constant)
    // ═══════════════════════════════════════════════════════════════

    // Kotlin: const val MAX_SIZE = 100
    // Equivalent Java: static final int MAX_SIZE = 100;
    // Must be: top-level or object property, primitive or String, no custom getter

    // ═══════════════════════════════════════════════════════════════
    // 6. LATEINIT (for deferred initialization)
    // ═══════════════════════════════════════════════════════════════

    // Kotlin has no "var" field without initial value (unlike Java)
    // Use lateinit for deferred initialization:
    lateinit var lateName: String
    lateName = "Initialized later"
    println("lateName: $lateName")

    // ═══════════════════════════════════════════════════════════════
    // 7. COMPARISON SUMMARY
    // ═══════════════════════════════════════════════════════════════
    //
    // Java                          →  Kotlin
    // ─────────────────────────────────────────────────
    // final int x = 5;              →  val x = 5
    // int x = 5;                    →  var x = 5
    // private final String s = "";  →  private val s = ""
    // private String s = "";        →  private var s = ""
    // static final int C = 10;     →  const val C = 10
    // String s; (late init)         →  lateinit var s: String
    //
    // KEY DIFFERENCES:
    // ─────────────────────────────────────────────────
    // 1. Java: var only for local vars. Kotlin: var everywhere
    // 2. Java: type BEFORE variable. Kotlin: type AFTER (val x: Int)
    // 3. Kotlin: const val for compile-time constants (not var!)
    // 4. Kotlin: lateinit for deferred init (Java: uninitialized field)
    // 5. val ≠ immutability of contents, only reference immutability
    //
    // ═══════════════════════════════════════════════════════════════
}
